wit/store.wit

```diff
@@ -0,0 +1,106 @@
+	/// depending on the specific implementation. For example:
+	///
+	/// 1. Amazon DynamoDB calls a collection of key-value pairs a table
+	/// 2. Redis has hashes, sets, and sorted sets as different types of collections
+	/// 3. Cassandra calls a collection of key-value pairs a column family
+	/// 4. MongoDB calls a collection of key-value pairs a collection
+	/// 5. Riak calls a collection of key-value pairs a bucket
+	/// 6. Memcached calls a collection of key-value pairs a slab
+	/// 7. Azure Cosmos DB calls a collection of key-value pairs a container
+	///
+	/// In this interface, we use the term `bucket` to refer to a collection of key-value pairs
+    resource bucket {
+        /// Get the store with the specified identifier.
+		///
+		/// `identifier` must refer to a bucket provided by the host.
+		///
+		/// `error::no-such-store` will be raised if the `identifier` is not recognized.
+        open: func(identifier: string) -> result<bucket, error>;
+
+        /// Get the value associated with the specified `key`
+        ///
+        /// The value is returned as an option. If the key-value pair exists in the
+        /// store, it returns `Ok(value)`. If the key does not exist in the
+        /// store, it returns `Ok(none)`. 
+        ///
+        /// If any other error occurs, it returns an `Err(error)`.
+        get: func(key: string) -> result<option<list<u8>>, error>;
+
+        /// Set the value associated with the key in the store. If the key already
+        /// exists in the store, it overwrites the value.
+        ///
+        /// If the key does not exist in the store, it creates a new key-value pair.
+        /// 
+        /// If any other error occurs, it returns an `Err(error)`.
+        set: func(key: string, value: list<u8>) -> result<_, error>;
+
+        /// Delete the key-value pair associated with the key in the store.
+        /// 
+        /// If the key does not exist in the store, it does nothing.
+        ///
+        /// If any other error occurs, it returns an `Err(error)`.
+        delete: func(key: string) -> result<_, error>;
+
+        /// Check if the key exists in the store.
+        /// 
+        /// If the key exists in the store, it returns `Ok(true)`. If the key does
+        /// not exist in the store, it returns `Ok(false)`.
+        /// 
+        /// If any other error occurs, it returns an `Err(error)`.
+        exists: func(key: string) -> result<bool, error>;
+
+        /// Get all the keys in the store. It returns a list of keys.
+        /// 
+        /// Note that the keys are not guaranteed to be returned in any particular order.
+        /// 
+        /// If the store is empty, it returns an empty list.
+        /// 
+        /// MAY show an out-of-date list of keys if there are concurrent writes to the store.
+        /// 
+        /// If any error occurs, it returns an `Err(error)`.
+        list-keys: func() -> result<list<string>, error>;
```

Comment:
Tell me if this was already discussed already, but is there a risk that a store has a gigantic number of keys and this OOMs?  In 0.3, we'd use a s/`list<string>`/`stream<string>`/, but in 0.2, we'd have to create a whole new pseudo-stream resource type (since `input-stream` returns bytes).  I don't know if this is a big enough problem in a 0.2.* timeframe to warrant all that work, but I thought I'd raise the question.

Reply:
That's a great point to raise! I've done some research and found that most key value stores either no do have direct ways to list all keys or provide a way with a warning. 

## Redis

You can use `KEYS` to list all the keys from Redis but it warns devleopers to not use this command in production cluster as it's known to be inefficient on large sets. 

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. If you're looking for a way to find keys in a subset of your keyspace, consider using [SCAN](https://redis.io/commands/scan) or [sets](https://redis.io/topics/data-types#sets).

The suggested `SCAN` command takes a PATTERN and COUNT and returns a cursor which can be used in the next invocation of `SCAN`, effectively implementing "paging" returns. 

## DynamoDB

DynamoDB doesn't have a direct equivalent to `list-keys` but it provides a `Scan` operation and sayd 

> If the total size of scanned items exceeds the maximum dataset size limit of 1 MB, the scan completes and results are returned to the user. The LastEvaluatedKey value is also returned and the requestor can use the LastEvaluatedKey to continue the scan in a subsequent operation.

It is also noted that this operation is resource intensive. 

## CosmosDB

CosmosDB does not provide a direct way to list keys, but you can do that with SQL that select all objects in a collection.

Reply:
Ah, great, thanks for listing the options.  It looks like large-number-of-keys is indeed a significant concern and at least the first two have a paging scheme that we would probably want to represent as a stream.  It does seem like we shouldn't attempt to return everything as a single `list`, but options in a 0.2 timeframe I think would be to (1) create a pseudo-stream (sitting in front of the cursor or paging API) of strings or (2) drop the feature for now until we can simply use `stream<string>`.

Reply:
What does that pseudo stream look like in practice? I get the general idea but not sure what it looks like with wit

Reply:
It would look rather like [`wasi:io/input-stream`](https://github.com/WebAssembly/wasi-io/blob/main/wit/streams.wit#L32), except swapping out `u8` with `string`.  In fact, maybe we'd want to add it to wasi-io as `string-stream` (so that any other "potentially-very-large list of strings" use cases could `use` it).

Reply:
I was honestly hoping to avoid streams until they were a bit easier to use. Would it be better to maybe implement the cursor/paging option or omit the `list-keys` as you suggested? @kate-goldenring curious on your thoughts here too

Reply:
I think support for getting keys is important for the interface, whether that be through a `scan-keys` function or a streamed `list-keys`. Another option would be to support `list-keys` with a warning (like with Redis) and leave it up to the host implementation to use paging when getting the keys.

Reply:
or `list-keys` with a `count` / `limit` parameter

Reply:
Yeah I think I am leaning towards a count parameter (that also gets returned with the response)

Reply:
Yes, that makes sense in a 0.2 timeframe.